# <center> Data Collection and Cleaning

**Overview** - The notebook serves the purpose of importing the data, converting the needed features to the desired datatypes, handling null values etc. <br>
**Result** - The cleaned data is the exported to be used further by the visualisation and prediction noteboo
**Process** - <br>
 - [Importing the Libraries](#Installing-Libraries)
 - [Importing the Data](#Importing-the-Data)
 - [Examining the Data](#Examining-the-Data)


## Installing Libraries


We will install the needfull libraries over here using the ```!pip install``` command.<br>
Libraries used are:<br>
- Pandas and Numpy : For data handling
- Seaborn and Matplotlib : For Visualisation
- Projects : To retrieve and add data to the project
`

This followed by importing the libraries for further use


In [5]:
!pip install pandas
!pip install numpy
!pip install seaborn
!pip install matplotlib

In [12]:
project_token = 'p-3f704f3cdf69a8a61cd7aa18710a153652a855a1'

In [13]:
import pandas as pd
import numpy as np
import seaborn as sns
import matplotlib.pyplot as plt
%matplotlib inline
from project_lib import Project
project = Project( project_access_token=project_token)
pc = project.project_context

## Importing the Data


The data is imported to the notebbok from the assets section of the project<br>
The code over here is imported directly from the watson data functionality

In [18]:
# The code was removed by Watson Studio for sharing.

We will install the needfull libraries over here using the ```.read_csv``` command of Pandas<br>
The imported csv is stored as a dtaframe with the name ```data```.

In [19]:
data = pd.read_csv(body)

Viewing the head of the data

In [20]:
data.head()

Source.Name web-scraper-order  \
0  cardekho_extract(0-2000).csv   1611917819-1662   
1  cardekho_extract(0-2000).csv   1611918361-1902   
2  cardekho_extract(0-2000).csv   1611917012-1306   
3  cardekho_extract(0-2000).csv   1611917695-1607   
4  cardekho_extract(0-2000).csv    1611914861-367   

                               web-scraper-start-url  \
0  https://www.cardekho.com/used-car-details/used...   
1  https://www.cardekho.com/used-car-details/used...   
2  https://www.cardekho.com/used-car-details/used...   
3  https://www.cardekho.com/used-car-details/used...   
4  https://www.cardekho.com/used-car-details/used...   

                                        full_name selling_price  \
0                                 Maruti Alto Std     1.2 Lakh*   
1                          Hyundai Grand i10 Asta     5.5 Lakh*   
2                                Hyundai i20 Asta    2.15 Lakh*   
3                   Maruti Alto K10 2010-2014 VXI    2.26 Lakh*   
4  Ford Ecosport 2015-2021 1.5 TDCi Titanium BSIV     5.7 Lakh*   

                                        new-price    year seller_type  \
0                                             NaN  2012.0  Individual   
1    New Car (On-Road Price) : Rs.7.11-7.48 Lakh*  2016.0  Individual   
2                                             NaN  2010.0  Individual   
3                                             NaN  2012.0  Individual   
4  New Car (On-Road Price) : Rs.10.14-13.79 Lakh*  2015.0      Dealer   

      km_driven   owner_type fuel_type transmission_type            mileage  \
0  1,20,000 kms  First Owner    Petrol            Manual   Mileage19.7 kmpl   
1    20,000 kms  First Owner    Petrol            Manual   Mileage18.9 kmpl   
2    60,000 kms  First Owner    Petrol            Manual   Mileage17.0 kmpl   
3    37,000 kms  First Owner    Petrol            Manual  Mileage20.92 kmpl   
4    30,000 kms  First Owner    Diesel            Manual  Mileage22.77 kmpl   

          engine           max_power   seats  
0   Engine796 CC   Max Power46.3 bhp  Seats5  
1  Engine1197 CC     Max Power82 bhp  Seats5  
2  Engine1197 CC     Max Power80 bhp  Seats5  
3   Engine998 CC   Max Power67.1 bhp  Seats5  
4  Engine1498 CC  Max Power98.59 bhp  Seats5

## Examining the Data

Now since the data is imported we will see the composition and various data types and look for null values etc.

Looking at the data shape

In [23]:
data.shape

(20026, 16)

So now we know that there are **20026** observations and **16** columns

Looking at data types of each column

In [24]:
data.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 20026 entries, 0 to 20025
Data columns (total 16 columns):
 #   Column                 Non-Null Count  Dtype  
---  ------                 --------------  -----  
 0   Source.Name            20026 non-null  object 
 1   web-scraper-order      20026 non-null  object 
 2   web-scraper-start-url  20026 non-null  object 
 3   full_name              19980 non-null  object 
 4   selling_price          19980 non-null  object 
 5   new-price              9566 non-null   object 
 6   year                   19980 non-null  float64
 7   seller_type            19980 non-null  object 
 8   km_driven              19980 non-null  object 
 9   owner_type             19980 non-null  object 
 10  fuel_type              19980 non-null  object 
 11  transmission_type      19980 non-null  object 
 12  mileage                19980 non-null  object 
 13  engine                 19921 non-null  object 
 14  max_power              19921 non-null  object 
 15  se

Null Values (absolute and percentage) by column
<br> This shows how many columns have a very high composition of null values rendering them useless

In [36]:
Column,NullValues,NullValuespercentage = [],[],[]
for i in data.columns:
    Column.append(i)
    NullValues.append(data[i].isnull().sum())
    NullValuespercentage.append(data[i].isnull().mean())
display(pd.DataFrame({'Column':Column,'Number of Null Values':NullValues,'% of null values':NullValuespercentage}).sort_values(by='% of null values',ascending=False).set_index('Column'))

Number of Null Values  % of null values
Column                                                        
new-price                              10460          0.522321
seats                                    173          0.008639
engine                                   105          0.005243
max_power                                105          0.005243
full_name                                 46          0.002297
selling_price                             46          0.002297
year                                      46          0.002297
seller_type                               46          0.002297
km_driven                                 46          0.002297
owner_type                                46          0.002297
fuel_type                                 46          0.002297
transmission_type                         46          0.002297
mileage                                   46          0.002297
Source.Name                                0          0.000000
web-scraper-order                          0          0.000000
web-scraper-start-url                      0          0.000000